index.js

```diff
@@ -58,6 +58,122 @@ class DbFacility extends Base {
       }
     ], cb)
   }
+
+  /**
+   * @see https://www.npmjs.com/package/mysql#pooling-connections
+   * @see https://www.npmjs.com/package/mysql#transactions
+   *
+   * @param {(conn: { cli: mysql.PoolConnection }, txFuncCb: (err: Error|undefined) => void) => void} func
+   * @param {(err: Error|undefined) => void} cb
+   */
+  runTransaction (func, cb) {
+    /** @type {mysql.PoolConnection} */
+    let conn = null
+    let txStarted = false
+    let txCommited = false
+
+    async.series({
+      conn: (next) => {
+        this.cli.getConnection((err, res) => {
+          if (err) return next(err)
+
+          conn = res // intentional
+          return next()
+        })
+      },
+      begin: (next) => {
+        conn.beginTransaction(next)
+      },
+      execute: (next) => {
+        txStarted = true
+        func({ cli: conn }, next)
+      },
+      commit: (next) => {
+        conn.commit(next)
+      },
+      release: (next) => {
+        txCommited = true
+        try {
+          conn.release()
+          return next()
+        } catch (err) {
+          return next(err)
```

Comment:
Since we only get to release after commit succeeds, why should failure to release the connection back to the pool shouldn't count as a transaction error? It looks like it'll lead to duplicate writes.

 Maybe the error at this stage should be ignored or we should actually merge the `release` task to the `commit` task. Something like this: 

```javascript
  commit: (next)=>{
     conn.commit(err=>{
         if(err) return next(err)
         txCommited = true
         try {  conn.release() }catch(err){ console.warn(err); next() }
     })
   }
```

Reply:
I guess something like this could help to have more control of state: https://github.com/bitfinexcom/bfx-facs-db-mysql/pull/2/files#r1102558676